src/components/ingestion/AbstractIngestionEngine.js

```diff
@@ -5,61 +5,90 @@ import {
+
 import logger from '%src/lib/logger'
 import kafka from '%src/lib/kafka'
+import { tsToSeconds } from '%src/lib/utils/mpegts'
 
 class AbstractIngestionEngine {
 	constructor() {
 		if (this.constructor === AbstractIngestionEngine) {
 			throw new AbstractInstantiationError(this.constructor.name)
 		}
 		this.producer = kafka.producer()
+		this.mpegtsDemuxer = new TSDemuxer(this.processDemuxedPacket)
+		this.ffmpegProcess = null
+		this.currentPts = 0
 	}
 
 	/**
-	 * Ingests data into the TV Kitchen pipeline.
+	 * Ingests a mpegts payload into the TV Kitchen pipeline.
 	 *
-	 * @param  {Buffer} data The buffer of data to be ingested.
-	 * @return {Boolean}     Whether or not the buffer was successfully ingested.
+	 * @param  {Payload} data The payload to be ingested.
+	 * @return {Boolean}      Whether or not the payload was successfully ingested.
 	 */
-	ingestData = async (data) => {
-		const payload = new Payload()
-		payload.data = data
+	ingestPayload = async (payload) => {
+		if (!(payload instanceof Payload)) {
+			throw new Error('ingestPayload must be passed a Payload object')
+		}
 		await this.producer.send({
-			topic: dataTypes.STREAM.DATA,
+			topic: dataTypes.STREAM.CONTAINER,
 			messages: [{
-				value: data,
+				value: Payload.serialize(payload),
 			}],
 		})
 		return true
 	}
 
 	/**
-	 * Continuously ingest of a stream of data.
+	 * Updates ingestion state based on the latest mpegts packet data as
+	 * parsed by our demuxer.
 	 *
-	 * @param  {ReadableStream} readableStream The stream of data to be recorded.
-	 * @return {ReadableStream}                The input stream (for chaining)
+	 * @param  {Packet} packet A Packet object, as defined by the TSDemuxer project
 	 */
-	recordStream = (readableStream) => readableStream
-		.on('data', async (data) => {
-			readableStream.pause()
-			await this.ingestData(data)
-			readableStream.resume()
+	processDemuxedPacket = (packet) => {
+		this.currentPts = packet.pts
+	}
+
+	/**
+	 * Process a new chunk of data from an mpegts stream.
+	 * This data is used to update local state and ultimately pipe to kafka.
+	 *
+	 * @param  {Buffer} mpegtsData The latest arbitrary chunk of buffered mpegts data.
+	 */
+	processMpegtsData = (mpegtsData) => {
+		this.ffmpegProcess.stdout.pause()
+		this.mpegtsDemuxer.process(mpegtsData)
+		const payload = new Payload({
+			data: mpegtsData,
+			ptsTime: tsToSeconds(this.currentPts),
 		})
+		this.ingestPayload(payload)
+		this.ffmpegProcess.stdout.resume()
+	}
 
 	/**
-	 * Start the ingestion engine.
+	 * Process a new chunk of raw data from the input stream.
+	 * This data is then piped to ffmpeg.
 	 *
-	 * @return {null}
+	 * @param  {Buffer} rawData The latest arbitrary chunk of raw input data.
+	 */
+	processRawData = (rawData) => {
+		this.inputStream.pause()
+		this.ffmpegProcess.stdin.write(rawData)
+		this.inputStream.resume()
+	}
+
+	/**
+	 * Start the ingestion engine.
+	 * This kicks off the ffmpeg stream and input stream pipeline.
```

Comment:
I think this second sentence could be expanded just a bit to make the order of operations more clear and descriptive. For instance, we intentionally:

1. Spawn an FFMpeg process to listen for input on stdin, then
2. Start the input stream and pipe it to the now-listening FFMpeg process

As-is, the sequential order of that is implicit when it would be helpful for the description to make it explicit.